File: Python/custom_functions.py
"""
Collection of custom functions produced in the development of the E-OBS-SWB2 project
Functions are divided in four sections:
0. Generic functions, useful in many contexts
1. E-OBS data handling
2. ArcASCII GRID creation (general)
3. netCDF input creation (SWB2 specific)
4. SWB2 netCDF output handling

@author: paolo
"""

# %% 0. Generic functions

def leap(y):
    #input: year (int)
    #output: number of days (int)
    if((y%4 == 0) | (y%400 == 0)):
        return 366
    else:
        return 365

def getkeys(dict):
    #Returns the dictionary keys of a dictonary as a list
    # got from:
    # https://www.geeksforgeeks.org/python-get-dictionary-keys-as-a-list/
    #Other method
    # return list(dict.keys())
    return [*dict]

def repeat_list(l, nrep, array = False):
    #Returns l repeated nrep times
    #Inspired from: https://stackoverflow.com/a/54483540
    import numpy as np
    rep = []
    for i in range(nrep):
        rep = rep + l
    rep = np.array(rep) if array else rep
    return rep

# %% 1. E-OBS data handling

def date_toeobs(y,m,d):
    #Returns how many days have passed from 1950-01-01 (starting counting day for E-OBS data)
    from datetime import date
    start = date(1950,1,1)
    d = date(y,m,d)
    count = d - start #how many days have passed since start
    return count.days

def eobs_todate(x, number = False):
    #Returns the year, month and day corresponding to the number given
    #If x is a plain number (not a variable), "number" must be set to True
    from datetime import date, timedelta
    start = date(1950,1,1)
    if (number):
        end = start + timedelta(days = x.item())
    else:
        end = start + timedelta(days = x)
    return end.year, end.strftime('%m'), end.strftime('%d')

def transf(lat_t, lon_t, zoneN, zoneL, var = 'xy'):
    #Transforms the lat/lon data provided in x/y projected UTM coordinates
    import utm
    x = []
    y = []
    for i in lat_t:
        for j in lon_t:
            #The syntax is utm.from_latlon(LATITUDE, LONGITUDE)
            #The return has the form (EASTING, NORTHING, ZONE_NUMBER, ZONE_LETTER)
            xj, yi, _, _ = utm.from_latlon(i, j,
                            force_zone_number = zoneN,
                            force_zone_letter = zoneL)
            x += [xj]
        y += [yi]
    x = x[0:len(lon_t)]
    if(var == 'xy'):
        return x, y
    elif(var == 'x'):
        return x
    elif(var == 'y'):
        return y

# %% 2. ArcASCII GRID creation (general)

def save_ArcGRID(df, fname, xll, yll, size, nodata):
    #df has to be a Pandas DataFrame
    #xll: x coordinate of the left bottom corner (lon)
    #yll: y coordinate of the left bottom corner (lat)
    #size: cell size (m)
    #nodata: value assigned to nodata
    def line_prepender(filename, line):
        with open(filename, 'r+') as f:
            content = f.read()
            f.seek(0, 0)
            f.write(line + '\n' + content)
            #line.rstrip('\r\n') if you want ot remove something from line
    df.to_csv(fname, sep = ' ', header = False, index = False)
    header = f'ncols         {len(df.columns)}\nnrows         {len(df.index)}\nxllcorner     {xll}\nyllcorner     {yll}\ncellsize      {size}\nNODATA_value  {nodata}'
    line_prepender(fname, header)

# %% 3. netCDF input creation (SWB2 specific)

def eobs_todaymet(y):
    #redefine: starting from 1980, adding 0.5
    from datetime import date
    dstart = date(1980, 1, 1)
    estart = date(1950, 1, 1)
    k = dstart - estart
    y1 = y - k.days + 0.5
    return y1

# %% 4. SWB2 netCDF output handling

def getdates(string, n):
    #Get the starting and ending dates of an SWB2 netCDF output file 
    from datetime import date, timedelta
    datestr = string.split(' ')[2]
    y, m, d = datestr.split('-')
    start = date(int(y), int(m), int(d))
    end = start + timedelta(days = n-1)
    return start, end

File: R/custom_functions.R
#Custom functions
# - General
# - SWB2 netCDF files handling

# General -----------------------------------------------------------------

save_ArcGRID <- function(df, fname, xll, yll, size, nodata){
  datafile <- file(fname, open = 'wt')
  on.exit(close(datafile))
  
  writeLines(paste0('ncols         ', dim(df)[1]), con = datafile)
  writeLines(paste0('nrows         ', dim(df)[2]), con = datafile)
  writeLines(paste0('xllcorner     ', xll), con = datafile)
  writeLines(paste0('yllcorner     ', yll), con = datafile)
  writeLines(paste0('cellsize      ', size), con = datafile)
  writeLines(paste0('NODATA_value  ', nodata), con = datafile)
  
  write(df, datafile, ncolumns = dim(df)[1], sep = ' ')
}

leap <- function(year){
  if(year %% 4 == 0 | year %% 400 == 0) return(366) else return(365)
}

# SWB2 netCDF handling ----------------------------------------------------


SPsum <- function(path, SPs, var,
                  starty, endy,
                  outpath = 'none'){
  #path: path to the netCDF file to be imported
  #SPs: stress periods lengths. Defined as cumulative sums
  library(ncdf4)
  
  print(paste0('Performing the sum of ', var, ' over the stress periods provided'))
  f = nc_open(path)
  varsize = f[["var"]][[var]][["size"]]
  if(outpath != 'none'){
    size = round(ncvar_get(f, 'x')[2] - ncvar_get(f, 'x')[1])
    xll = ncvar_get(f, 'x')[1] - size/2
    yll = ncvar_get(f, 'y')[length(ncvar_get(f, 'y'))] - size/2
  }
  period = seq(starty, endy)
  s = k = 1
  var3d = array(0, dim = c(length(period)*length(SPs), varsize[1], varsize[2]))
  for (y in period){
    year = ncvar_get(f, var, start = c(1, 1, s), count = c(varsize[1], varsize[2], leap(y)))
    base = 1
    for (i in seq_len(length(SPs))){
      SP = ifelse((leap(y) == 366), SPs[i]+1, SPs[i])
      sp = year[, , base:SP]
      base = SP+1
      if((leap(y) == 366) & (i == 1)){
        print(paste0('Be careful, when transforming to m/s, consider that SP1 of year ', y, ' has ', dim(sp)[3], ' days'))
      }
      for(j in seq_len(dim(sp)[3])) var3d[k, , ] = var3d[k, , ] + sp[, , j]
      if (outpath != 'none'){
        fname = paste0(outpath, '/', var, '_', y, '_SP', i, '.asc')
        save_ArcGRID(var3d[k, , ], fname, xll, yll, size, -9999)
      }
      k = k + 1
    }
    s = s + leap(y)
  }
  if (outpath != 'none') print(paste0('The ASCII files are saved in ', outpath))
  nc_close(f)
  return(var3d)
}
